### DISCUSSION TITLE
Opinionated Eruption

### DISCUSSION THREAD

**Discussion by raisiqueira** [eruptionjs/core#36] [Ideas]
The main idea with an opinionated template is to provide a template (based on [Eruption/core](https://github.com/eruptionjs/core)) with a set of libraries that help the developer to start a new project without so much pain.

e.g., of libraries: 

- React Router
- React Query (?)

**Comment by luigieai** [comment]
Well, we may need to define types of library that we can suggest for the user, like:

- Test libs
- Dev libs
- React libs
- Astro libs (when we start supporting)

And define what libraries should be best for our opiniated boilerplate

**Comment by devmozao** [comment]
As of today, we provide the following on Eruption: engines, deps and dev deps kind of libs. 

Engines ones run the core of Eruption itself. They are `nodejs`, `npm`, `vite` and `typescript`. We must have them and in a short future, maybe even offer alternatives to them, once bun/deno starts being a reliable and easy-to-code offer.

Dependecies are also the core of Eruption today, since we still don't support other frameworks, yet 👀. They are `react` and `react-dom`. 

And then there is Devs Dependency. Those are critical to make Eruption being... well, Eruption. It's on the shoulder of these giants that we create the kind of DX we want for Eruption. We need all of them, ATM. Will avoid going every single one of them, but a list can be found in [core/package.json](https://github.com/eruptionjs/core/blob/b5d3ca864ef5f28f104dae5b082f852711efcb12/package.json#L39-L73)

That being said, I'll focus on what libs we will opt in, in order to make the the opinionated version and why I'm voting them. So let's go: 

**project compression:** [vite-plugin-compression2](https://github.com/nonzzz/vite-compression-plugin)
need to define if this will be opinionated or default, but I think it's also need further investigation (see [this](https://github.com/alloc/vite-plugin-compress)).

**for routes:** [react-router-dom](https://reactrouter.com/en/main)
I currently having dificulties with the 6.9 at work, but i'll test case a new approach of how to use it for our needs at work, so probably what works there will be my suggestion here

**requests:** [react-query](https://tanstack.com/query/v3/) with [axios](https://github.com/axios/axios)
interceptors and easy to type request/response. I also have a current mental model for that in my work of how we're using, but we can discuss later and even being an option inside the full version

**css/components:** [mantine/core](https://mantine.dev) with mantine/hooks and emotion
Today, it's the best documented components lib out there. Friendly to work with typescript, and easy to create custom components, and also change default component configs. Our experience in our project, is that it's beating MUI like 10 - 0, and every single way.

Also I feel here that currently, tailwind is out of scope for a highly opinionated project. It's cool for smallers ones but besides of that, the crazyness that is those class creates, it's just a big NOPE to medium and bigger projects.

**icons:** to be defined
Currently I don't have a good answer for that. Material-icons was a good and reliable source, but Google screwed up, as they usually do: merged back the icons with the ui, makint it impossible to use v5 withouth the whole project. (and other alternatives related to material-icon sucks atm). Maybe go back to fontawesome? They have a free tier now. Our project is using tabler/icons-react and it's nice some features like stroke and size, but they lack of finish in fine details like exact padding/margin as default for centering, and the design is slightly worst than material-icons.

**font:** [inter-ui](https://github.com/rsms/inter)
it's open source and goes well in a theme config with mantine.

And I think that's it. (for now) what do you guys think?

PS: for "we", I mean my team at Makasí working on the main product

**Comment by luigieai** [comment]
As I'm inexperienced with frontend development and I'm more focused at CLI and automations tasks, I don't have anything to disagree with, as for Icons, I can suggest [css.gg](https://css.gg/) as it's open source, have some compatibility with Figma, and is easy to implement.

**Comment by raisiqueira** [comment]
Hey folks, I started a PoC under this repository: https://github.com/raisiqueira/opinionated-eruption

- Added React Query
- Added Mantine as component library

To-do list:

- Add React Router and integrate it with React Query.
- Add a real world example of how to use React Query (with tests)

I'll try to finish it until tomorrow, and we could discuss folder structure, and other architecture points.

**Reply by devmozao** [reply]
I need to show you the craziest thing that I'm doing with root imports with folder structure. It's like, mental, but it's really easy to grasp and make people productive. ATM I have a new jr dev who is in her first job as a dev, and in the first week she was already making features and contribution, because its easy to get going.

Anyway, i'll share soon and the why's.

**Comment by kevenleone** [comment]
Hey guys, just jumping into discussion here 👋🏻 

For Data Fetching [SWR](https://swr.vercel.app/pt-BR) is a good option too, it works similar to React Query and it's really simple.

```js
import useSWR from 'swr'
 
function Profile() {
  const { data, error, isLoading } = useSWR('/api/user', fetcher)
 
  if (error) return <div>falhou em carregar</div>
  if (isLoading) return <div>carregando...</div>
  return <div>Olá {data.name}!</div>
}
```

Maybe something to consider is to provide a simple way to handle with forms, as @raisiqueira said:
> "start a new project without so much pain."

Can't find a pain higher than that, if it makes sense, it could be viable to consider:
Form Libraries: **React-Hook-Form** or **Formik**  for validation **Zod** or **Yup**.

-----

Options for icons:

* https://lucide.dev
* https://react-icons.github.io/react-icons

**Reply by raisiqueira** [reply]
Thanks for the contribution, @kevenleone. I believe that RHF and Zod is a good combination, I'll add it on my PoC. About SWR over React Query, I think that React Query is a great option because IMO there's more content on internet, such as tutorials, how to solve common issues, etc.

**Reply by devmozao** [reply]
Interesting. I do agree that RHF and Zod is the way to solve forms. (not that I don't like Formik with Yup. Used in some past projects, but RHF have a better performance, better docs and better dx).

About SWR vs RQ: I think it's an interesting approaching. Maybe could be one of the options that we could provide to the CLI when we implement the proccess where the developer can choose all options to his project, so then, he could choose like, "do I want pure SWC or RQ with axios, or RQ with fetch?".

The way that I see is, that there more points that make each of this decision such a relevant one, like: do we want retry on errors? do we want manual refetch? sync between mutations and queries? do we want interceptors? and an even more relevant: do we want cache as default? All of these questions weights heavily to choose what solution is the best. At this moment, RQ/axios is full implementation of those for the opinionated one, but I believe that we're evolving to a scenario and necessity that the CLI will be way much more of a tool that you chooses your boilerplate only, in the sence that we can develop a tool that even after the initial boilerplate, we will still use to pre-config/install others patterns on the fly. But we will get there, eventually. Baby steps.

**Comment by devmozao** [comment]
it took some time, but here it is: https://github.com/eruptionjs/core/pull/59
